LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/Ipmi.cs

```diff
@@ -0,0 +1,410 @@
+
+    private ManagementObject _ipmi;
+
+    const byte COMMAND_GET_SDR_REPOSITORY_INFO = 0x20;
+    const byte COMMAND_RESERVE_SDR_REPOSITORY = 0x22;
+    const byte COMMAND_GET_SDR = 0x23;
+    const byte COMMAND_GET_SENSOR_READING = 0x2d;
+
+    const byte COMMAND_FAN_MODE = 0x45;
+    const byte COMMAND_FAN_LEVEL = 0x70;
+
+    const byte FAN_MODE_FULL = 0x01;
+    const byte FAN_MODE_OPTIMAL = 0x02;
+
+    const byte NETWORK_FUNCTION_SENSOR_EVENT = 0x04;
+    const byte NETWORK_FUNCTION_STORAGE = 0x0a;
+    const byte NETWORK_FUNCTION_SUPERMICRO = 0x30;
+
+    private bool _touchedFans = false;
+
+    public Ipmi(Manufacturer motherboardManufacturer)
+    {
+        Chip = Chip.IPMI;
+        _motherboardManufacturer = motherboardManufacturer;
+
+        ManagementClass ipmiClass = new ManagementClass("root\\WMI", "Microsoft_IPMI", null);
+
+        foreach (ManagementObject ipmi in ipmiClass.GetInstances())
+        {
+            _ipmi = ipmi;
+        }
+
+        // Fan control is exposed for Supermicro only as it differs between IPMI implementations
+        if (_motherboardManufacturer == Manufacturer.Supermicro)
+        {
+            ControlNames.Add("CPU Fan");
+            ControlNames.Add("System Fan");
+        }
+
+        // Perform an early update to count the number of sensors and get their names
+        Update();
+
+        Controls = new float?[_controls.Count];
+        Fans = new float?[_fans.Count];
+        Temperatures = new float?[_temperatures.Count];
+        Voltages = new float?[_voltages.Count];
+    }
+
+    public string GetReport()
+    {
+        StringBuilder sb = new();
+        Update(sb);
+        return sb.ToString();
+    }
+
+    public void SetControl(int index, byte? value)
+    {
+        if (_motherboardManufacturer == Manufacturer.Supermicro)
+        {
+            if (value != null || _touchedFans)
+            {
+                _touchedFans = true;
+
+                if (value == null)
+                {
+                    RunIPMICommand(COMMAND_FAN_MODE, NETWORK_FUNCTION_SUPERMICRO, new byte[] { 0x01 /* Set */, FAN_MODE_OPTIMAL });
+                }
+                else
+                {
+                    byte[] fanMode = RunIPMICommand(COMMAND_FAN_MODE, NETWORK_FUNCTION_SUPERMICRO, new byte[] { 0x00 });
+                    if (fanMode.Length < 2 || fanMode[0] != 0 || fanMode[1] != FAN_MODE_FULL)
+                    {
+                        RunIPMICommand(COMMAND_FAN_MODE, NETWORK_FUNCTION_SUPERMICRO, new byte[] { 0x01 /* Set */, FAN_MODE_FULL });
+                    }
+
+                    float speed = (float)value / 255.0f * 100.0f;
+                    RunIPMICommand(COMMAND_FAN_LEVEL, NETWORK_FUNCTION_SUPERMICRO, new byte[] { 0x66, 0x01 /* Set */, (byte)index, (byte)speed });
+                }
+            }
+        }
+        else
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    public void Update() { Update(null); }
```

Comment:
Can you clean this document up / reformat to follow code style?

Reply:
Have reformatted a bit, let me know if anything else needs tweaking